src/cmd/services/m3coordinator/downsample/metrics_appender.go

```diff
@@ -242,13 +244,17 @@ func (a *metricsAppender) SamplesAppender(opts SampleAppenderOptions) (SamplesAp
 	// N.B - it's safe the reuse the shared tag iterator because the matcher uses it immediately to resolve the optional
 	// namespace tag to determine the ruleset for the namespace. then the ruleset matcher reuses the tag iterator for
 	// every match computation.
-	matchResult, err := a.matcher.ForwardMatch(a.tagIter, fromNanos, toNanos, rules.MatchOptions{
+	matchOptions := &rules.MatchOptions{
 		NameAndTagsFn:       a.nameTagFn,
 		SortedTagIteratorFn: a.tagIterFn,
-	})
-	if err != nil {
+		MatchResult:         &a.matchResult,
+	}
+	if err := a.matcher.ForwardMatch(a.tagIter, fromNanos, toNanos, matchOptions); err != nil {
 		return SamplesAppenderResult{}, err
 	}
+	// Use the MatchResult on matchOptions as opposed to matchResult member variable
+	// as certain matchers may replace the MatchResult (e.g. cache matcher).
+	matchResult := matchOptions.MatchResult
```

Comment:
i don't think we want to allow this, since it defeats the entire purpose of the change. I think we'll need to restructure the cache matcher to use the provided results.